Source/Dafny/Resolver.cs

```diff
@@ -5995,6 +5995,10 @@ enum TailRecursionStatus
         }
         return status;
       } else if (stmt is AssignSuchThatStmt) {
+      } else if (stmt is AssignOrReturnStmt) {
+        // TODO this should be the conservative choice, but probably we can consider this to be tail-recursive
+        // under some conditions? However, how does this interact with compiling to exceptions?
+        return TailRecursionStatus.NotTailRecursive;
```

Comment:
I think this is the right choice. For it to be tail recursive, the `:-` call would have to be followed by something that wraps the `:-` LHS into a Success value assigned to the out-parameter. That sounds like a complicated check to me.